spark/src/main/scala/geotrellis/spark/ContextRDDMethods.scala

```diff
@@ -13,7 +13,7 @@ class ContextRDDMethods[K: ClassTag, V: ClassTag, M](val rdd: RDD[(K, V)] with M
   def mapKeys[R: ClassTag](f: K => R): RDD[(R, V)] with Metadata[M] =
     rdd.withContext { rdd => rdd.map { case (key, tile) => f(key) -> tile } }
 
-  def mapValues(f: V => V): RDD[(K, V)] with Metadata[M] =
+  def mapValues[W](f: V => W): RDD[(K, W)] with Metadata[M] =
```

Comment:
Good catch, this should have better documented. Here is the use-case:

Given instance of  `RDD[(K, V)] with Metadata[M]`
call to `.mapValues` will always resolve to `ContextRDDMethods.mapValues` as it is the most specific implicit class. Scala will not go a-looking elsewhere for the function passed in changes the value type.

Which means that if a user is intending to use `.mapValues` in the way that spark defines it, with mapping `V => W`, it will fail with something like: 
```scala
[error]  found   : (Int, None.type)
[error]  required: Int
[error]     val expected = pr1.mapValues(v => (v, None))
```

I think we talked about enforcing `.mapValues(f: V => V)` type signature and we can still do it if we change the name. However, it’s not clear how much this is buying us. Let me argue against that goal.

Keeping the type of values is not guarantee that metadata is valid. For instance `.convert` function will change the content of the tile but not the type. So we have to do some book keeping and adjust the metadata with it. https://github.com/echeipesh/geotrellis/blob/feature/partitioner/spark/src/main/scala/geotrellis/spark/BaseRasterRDDMethods.scala#L7-L8

Some of the problem stems from the fact that even now part of the metadata, in case of `RasterMetaData`, refers to the keys (`extent`, `layout`, `crs`) and part refers to values (`cellType`). In future we may have `bandCount` which also refers to values only.

One way to deal with this is to maybe to provide two traits to be extended by metadata, one for keys and one for values and provide API that changes one or the other depending on the function. If our objective remains to support user extensible metadata we basically have no control of what that might be and would have to rely on lenses most likely.

Another way to look at this is as user concern. In this case we provide a way to mutate metadata with the keys and/or values and rely on the user to be a good programmer. Our library code can again rely on lenses but this time lenses for the whole metadata type. For instance to change layout during pyramiding (https://github.com/echeipesh/geotrellis/commit/4d252901427d6337175accb42535d9814f92422d).

I’m starting to be in favor of the second approach as it more accurately reflects the reality, metadata is liable to go invalid without care. Plus splitting the metadata type into key and value metadata doesn’t seem to buy much as both approaches would fall-back on lenses.